internal/allocation/allocation.go

```diff
@@ -52,8 +52,9 @@ func addr2IPFingerprint(addr net.Addr) string {
 }
 
 // NewAllocation creates a new instance of NewAllocation.
-func NewAllocation(turnSocket net.PacketConn, fiveTuple *FiveTuple, log logging.LeveledLogger) *Allocation {
+func NewAllocation(turnSocket net.PacketConn, clientAddr net.Addr, fiveTuple *FiveTuple, log logging.LeveledLogger) *Allocation {
```

Comment:
Whats your motivation to pass the `clientAddr` separately instead of taking it from `fiveTuple`?

I think we could add some getters to the `FiveTuple` which provides a `net.{TCP,UDP}Addr` depending on `FiveTuple.Protocol`.

Reply:
Here is an example:

```go
// SrcAddr returns the tuples source address as net.Addr
func (t FiveTuple) SrcAddr() net.Addr {
	switch t.protocol {
	case UDP:
		return &net.UDPAddr{
			IP:   t.srcAddr.Addr().AsSlice(),
			Port: int(t.srcAddr.Port()),
		}
	case TCP:
		return &net.TCPAddr{
			IP:   t.srcAddr.Addr().AsSlice(),
			Port: int(t.srcAddr.Port()),
		}
	default:
		return nil
	}
}
```

Reply:
This was *exactly* the approach I implemented first but then it occurred to me that this way we would do *two* address type conversions per each packet: first from `net.UDPAddr`/`net.TCPAddr` to `netip.AddrPort` when we build the `FiveTuple` to call `GetAllocation`, and then back again to `net.UDPAddr`/`net.TCPAddr` in the `SrcAddr` function to pass the definition address to`WriteTo`. Now I don't know how costly these conversions are but I thought that if we can get away with just one then this would buy us some extra performance, and storing the original source address as a `net.UDPAddr`/`net.TCPAddr` in the allocation allows us to do that.

That being said, your code is still significantly simpler so if you insist I'll be happy to implement it. Wdyt?

Reply:
Good question. I am not sure about it.. It could be a case of premature optimization. So to have a conclusive decision we should probably benchmark it... But in general I would prefer readability over performance if we can not show the performance benefits.

Reply:
I think if readability is number one priority (I can relate to this POV) then we should drop this PR all together. It already decreases readability by a large margin and I could not show massive performance improvement. I don't say there's no improvement (we remove some 4-7% overhead from `GetAllocation`), all I'm saying is that the end-to-end tests do not seem to show significant speedup. 

Conclusion: we are I/O bounded, so optimizing code will not help...